packages/deployments/contracts/contracts/facets/StableSwapFacet.sol

```diff
@@ -0,0 +1,458 @@
+    if (s.swapStorages[_canonicalId].pooledTokens.length != 0)
+      revert StableSwapFacet__initializeSwap_alreadyInitialized();
+
+    // Check _pooledTokens and precisions parameter
+    if (_pooledTokens.length <= 1 || _pooledTokens.length > 32)
+      revert StableSwapFacet__initializeSwap_invalidPooledTokens();
+
+    if (_pooledTokens.length != decimals.length) revert StableSwapFacet__initializeSwap_decimalsMismatch();
+
+    uint256[] memory precisionMultipliers = new uint256[](decimals.length);
+
+    for (uint8 i = 0; i < _pooledTokens.length; i++) {
+      if (i > 0) {
+        // Check if index is already used. Check if 0th element is a duplicate.
+        if (s.tokenIndexes[_canonicalId][address(_pooledTokens[i])] != 0 || _pooledTokens[0] == _pooledTokens[i])
+          revert StableSwapFacet__initializeSwap_duplicateTokens();
+      }
+      if (address(_pooledTokens[i]) == address(0)) revert StableSwapFacet__initializeSwap_zeroTokenAddress();
+
+      if (decimals[i] > SwapUtils.POOL_PRECISION_DECIMALS)
+        revert StableSwapFacet__initializeSwap_tokenDecimalsExceedMax();
+
+      precisionMultipliers[i] = 10**uint256(SwapUtils.POOL_PRECISION_DECIMALS - decimals[i]);
+      s.tokenIndexes[_canonicalId][address(_pooledTokens[i])] = i;
+    }
+
+    // Check _a, _fee, _adminFee, _withdrawFee parameters
+    if (_a >= AmplificationUtils.MAX_A) revert StableSwapFacet__initializeSwap_aExceedMax();
+    if (_fee >= SwapUtils.MAX_SWAP_FEE) revert StableSwapFacet__initializeSwap_feeExceedMax();
+    if (_adminFee >= SwapUtils.MAX_ADMIN_FEE) revert StableSwapFacet__initializeSwap_adminFeeExceedMax();
+
+    // Initialize a LPToken contract
+    LPToken lpToken = LPToken(Clones.clone(lpTokenTargetAddress));
+    if (!lpToken.initialize(lpTokenName, lpTokenSymbol))
+      revert StableSwapFacet__initializeSwap_failedInitLpTokenClone();
+
+    // Initialize swapStorage struct
+    s.swapStorages[_canonicalId] = SwapUtils.Swap({
+      initialA: _a * AmplificationUtils.A_PRECISION,
+      futureA: _a * AmplificationUtils.A_PRECISION,
+      swapFee: _fee,
+      adminFee: _adminFee,
+      lpToken: lpToken,
+      pooledTokens: _pooledTokens,
+      tokenPrecisionMultipliers: precisionMultipliers,
+      balances: new uint256[](_pooledTokens.length),
+      initialATime: 0,
+      futureATime: 0
+    });
+  }
+
+  /**
+   * @notice Withdraw all admin fees to the contract owner
+   * @param canonicalId the canonical token id
+   */
+  function withdrawAdminFees(bytes32 canonicalId) external onlyOwner {
+    s.swapStorages[canonicalId].withdrawAdminFees(msg.sender);
+  }
+
+  /**
+   * @notice Update the admin fee. Admin fee takes portion of the swap fee.
+   * @param canonicalId the canonical token id
+   * @param newAdminFee new admin fee to be applied on future transactions
+   */
+  function setAdminFee(bytes32 canonicalId, uint256 newAdminFee) external onlyOwner {
+    s.swapStorages[canonicalId].setAdminFee(newAdminFee);
```

Comment:
imo, should have a MAX_ADMIN_FEE and a MAX_SWAP_FEE in place and require that this `newAdminFee` is always <= to it.

MAX_SWAP_FEE should be like 1% or 100 basis points maybe (so 10**8?)

Admin fee technically does not take any fee on top of swap fee, it takes a % of the swap fee iirc. So it should be capped at like, 50%? Does that sound right?

Just want to add rug protection for users here in case of unlikely event owner is compromised lol